design/validation-spi.md

```diff
@@ -0,0 +1,293 @@
+# Validation SPI
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-2045](https://issues.redhat.com/browse/KEYCLOAK-2045)
+
+## Motivation
+
+Validation is currently scattered around several places and implemented in different ways 
+which makes it very hard to reuse validation logic in a consistent way. 
+Besides some static helper functions that are used in different places, there are also many 
+component-specific validations that are directly wired to the surrounding business logic.
+A unified and consistent validation mechanism is currently missing.
+
+There have been several approaches to establish a validation infrastructure, but they have 
+not been followed consistently.
+
+The goal of this proposal is to provide an extensible and unified validation SPI 
+that can be used across the Keycloak code-base. The need for a new validation API was highlighted 
+by the design discussions for the [user-profile](https://github.com/keycloak/keycloak-community/blob/master/design/user-profile.md) support.
+
+The goals of this proposal can be summarized as:
+
+* Provide a uniform and extensible validation SPI
+* Consolidate the scattered validation logic with the new Validation SPI
+
+## Design Principles
+
+The design of the new validation SPI should follow the following design principles:
+
+- Simplicity
+The API should have a small API surface and be easy to use.
+
+- API Granularity
+The API should support low-level validations.
+
+- Efficiency
+Validations should be fast and have little overhead.
+
+- Extensibility
+There should be support for built-in as well as user provided validations.
+The new SPI should be easy to extend, most custom validations should only need one implementation class. 
+Developers should be able to provide reusable custom validation logic components.
+
+- Versatility
+Validations should be flexible and should support validatation of simple values and complex objects.
+It should be possible to parameterize validation executions. Additionally there should be a way to 
+check if the parameterization for a validation is valid.
+Validations should be composable and it should be possible to access other validations from with a validation. 
+
+## Considered Options
+
+### Bean Validation API
+
+The JSR 380 Bean Validation 2.0 is a mature and rich validation framework that is extensively used in Java applications.
+
+- Optimized for validation of complex Java bean objects based on declarative configuration rather than single values.
+
+Allthough, value validations are possible via `javax.validation.Validator#validateValue`, they require a bean class 
+with a property with the annotated constraints.
```

Comment:
That's nice and those programmatic constraint definitions can be created from a `UserProfile` config, but how does this work with the dynamic nature of `UserProfile`? There are no explicit fields or methods that expose the user profile properties, since it's effectively backed by a `Map` of attributes. Is there a way to plug-in your own property resolvers?